cudax/include/cuda/experimental/__container/heterogeneous_iterator.cuh

```diff
@@ -0,0 +1,430 @@
+//===----------------------------------------------------------------------===//
+//
+// Part of CUDA Experimental in CUDA C++ Core Libraries,
+// under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+// SPDX-FileCopyrightText: Copyright (c) 2024 NVIDIA CORPORATION & AFFILIATES.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef __CUDAX__CONTAINERS_HETEROGENEOUS_ITERATOR_CUH
+#define __CUDAX__CONTAINERS_HETEROGENEOUS_ITERATOR_CUH
+
+#include <cuda/std/detail/__config>
+
+#if defined(_CCCL_IMPLICIT_SYSTEM_HEADER_GCC)
+#  pragma GCC system_header
+#elif defined(_CCCL_IMPLICIT_SYSTEM_HEADER_CLANG)
+#  pragma clang system_header
+#elif defined(_CCCL_IMPLICIT_SYSTEM_HEADER_MSVC)
+#  pragma system_header
+#endif // no system header
+
+#include <cuda/__memory_resource/properties.h>
+#include <cuda/std/__iterator/iterator_traits.h>
+#include <cuda/std/__memory/addressof.h>
+#include <cuda/std/__memory/pointer_traits.h>
+#include <cuda/std/__type_traits/maybe_const.h>
+#include <cuda/std/cstdint>
+
+#include <cuda/experimental/__utility/select_execution_space.cuh>
+
+#if defined(LIBCUDACXX_ENABLE_EXPERIMENTAL_MEMORY_RESOURCE)
+
+//! @file The \c heterogeneous_iterator class is an iterator that provides typed execution space safety.
+namespace cuda::experimental
+{
+//! @rst
+//! .. _cudax-containers-heterogeneous-iterator:
+//!
+//! Type safe iterator over heterogeneous memory
+//! ---------------------------------------------
+//!
+//! ``heterogeneous_iterator`` provides a type safe access over heterogeneous memory. Depending on whether the memory is
+//! tagged as host-accessible and / or device-accessible the iterator restricts memory access.
+//! All operations that do not require memory access are always available on host and device.
+//!
+//! @endrst
+//! @tparam _Tp The underlying type of the elements the \c heterogeneous_iterator points at.
+//! @tparam _IsConst Boolean, if false the \c heterogeneous_iterator allows mutating the element pointed to.
+//! @tparam _Properties The properties that the \c heterogeneous_iterator is tagged with.
+template <class _Tp, bool _IsConst, class... _Properties>
+class heterogeneous_iterator;
```

Comment:
how about `heterogeneous_iterator<T const, Properties...>` instead of `heterogeneous_iterator<T, true, Properties...>`? boolean template parameters don't read well.

Reply:
That does not work because the underlying value type of the iterator must still be `T` I agree that moving towards an enum would be better

Reply:
> That does not work because the underlying value type of the iterator must still be `T`

`value_type` can still be `T` if you define it like:

```c++
using value_type = remove_const_t<T>;
```

Reply:
Changed